File: lib/thermalIndex.ts
/**
 * KIRAN — Thermal Stress Index (TSI)
 * -----------------------------------
 * IMPORTANT: This function must stay byte-for-byte identical in logic to
 * backend/thermal_index.py. The frontend NEVER recomputes this — it only
 * displays what the backend already calculated and stored. This copy exists
 * here only for local dev/testing/storybook use, not for production display.
 *
 * Base method: Rothfusz Heat Index regression, adjusted with an
 * India-calibration offset (Indian populations show heat-stress onset at
 * slightly higher humidity than the temperate-climate-derived NWS formula
 * assumes). Thresholds are mapped to IMD/NDMA Heat Action Plan tiers:
 *   Yellow  -> Low / Moderate
 *   Orange  -> High
 *   Red     -> Extreme
 */

export type RiskCategory = "Low" | "Moderate" | "High" | "Extreme";

export interface ThermalInput {
  tempC: number; // dry bulb air temperature, Celsius
  rh: number; // relative humidity, 0-100
  windKmh?: number; // wind speed, km/h (optional refinement)
}

export interface ThermalResult {
  score: number; // normalized 0-100 stress score
  heatIndexC: number; // apparent temperature, Celsius
  category: RiskCategory;
  ndmaTier: "Yellow" | "Orange" | "Red";
}

const INDIA_HUMIDITY_TOLERANCE_OFFSET = 0.92; // dampens RH contribution vs NWS baseline

function toFahrenheit(c: number) {
  return c * 9 / 5 + 32;
}
function toCelsius(f: number) {
  return (f - 32) * 5 / 9;
}

export function calculateThermalStress(input: ThermalInput): ThermalResult {
  const { tempC, rh, windKmh = 0 } = input;
  const T = toFahrenheit(tempC);
  const R = rh * INDIA_HUMIDITY_TOLERANCE_OFFSET;

  // Rothfusz regression (NOAA), valid for T >= 80F; simple average used below it
  let hiF: number;
  if (T < 80) {
    hiF = 0.5 * (T + 61.0 + (T - 68.0) * 1.2 + R * 0.094);
  } else {
    hiF =
      -42.379 +
      2.04901523 * T +
      10.14333127 * R -
      0.22475541 * T * R -
      0.00683783 * T * T -
      0.05481717 * R * R +
      0.00122874 * T * T * R +
      0.00085282 * T * R * R -
      0.00000199 * T * T * R * R;
  }

  // Wind speed provides mild evaporative relief above 8 km/h
  if (windKmh > 8) {
    hiF -= Math.min((windKmh - 8) * 0.05, 3);
  }

  const heatIndexC = Math.round(toCelsius(hiF) * 10) / 10;

  // Map heat index (C) to a 0-100 score, calibrated so that:
  // <30C -> 0-40 (Low), 30-38C -> 40-60 (Moderate),
  // 38-45C -> 60-80 (High), >45C -> 80-100 (Extreme)
  // matching IMD heatwave departure criteria for the Indian plains.
  let score: number;
  if (heatIndexC <= 30) {
    score = Math.max(0, (heatIndexC / 30) * 40);
  } else if (heatIndexC <= 38) {
    score = 40 + ((heatIndexC - 30) / 8) * 20;
  } else if (heatIndexC <= 45) {
    score = 60 + ((heatIndexC - 38) / 7) * 20;
  } else {
    score = Math.min(100, 80 + ((heatIndexC - 45) / 5) * 20);
  }
  score = Math.round(score * 10) / 10;

  let category: RiskCategory;
  let ndmaTier: "Yellow" | "Orange" | "Red";
  if (score < 40) {
    category = "Low";
    ndmaTier = "Yellow";
  } else if (score < 60) {
    category = "Moderate";
    ndmaTier = "Yellow";
  } else if (score < 80) {
    category = "High";
    ndmaTier = "Orange";
  } else {
    category = "Extreme";
    ndmaTier = "Red";
  }

  return { score, heatIndexC, category, ndmaTier };
}

export const riskColor: Record<RiskCategory, string> = {
  Low: "#2DD4BF",
  Moderate: "#F59E0B",
  High: "#FB6D3A",
  Extreme: "#EF4444",
};

export const sunCompanionMessage: Record<RiskCategory, { en: string; hi: string }> = {
  Low: { en: "I'm feeling normal today.", hi: "आज मौसम सामान्य है।" },
  Moderate: { en: "It's getting warm out there.", hi: "बाहर गर्मी बढ़ रही है।" },
  High: { en: "It's quite hot — take care.", hi: "काफी गर्मी है — ध्यान रखें।" },
  Extreme: { en: "It's extremely hot — please stay safe.", hi: "अत्यधिक गर्मी है — कृपया सुरक्षित रहें।" },
};

File: backend/thermal_index.py
"""
KIRAN - Thermal Stress Index (TSI)
-----------------------------------
This is THE single source of truth for the Thermal Stress Index calculation.
Call this exact function from:
  - the Supabase Edge Function that writes new rows into `thermal_index`
  - the FastAPI model service (if you run a separate ML/forecast service)
Never reimplement this logic anywhere else, including the frontend.

Base method: Rothfusz Heat Index regression (NOAA), adjusted with an
India-calibration offset, mapped to IMD/NDMA Heat Action Plan tiers.
"""

from dataclasses import dataclass
from typing import Literal

RiskCategory = Literal["Low", "Moderate", "High", "Extreme"]
NDMATier = Literal["Yellow", "Orange", "Red"]

INDIA_HUMIDITY_TOLERANCE_OFFSET = 0.92  # dampens RH contribution vs NWS baseline


@dataclass
class ThermalResult:
    score: float          # normalized 0-100 stress score
    heat_index_c: float   # apparent temperature, Celsius
    category: RiskCategory
    ndma_tier: NDMATier


def _to_f(c: float) -> float:
    return c * 9 / 5 + 32


def _to_c(f: float) -> float:
    return (f - 32) * 5 / 9


def calculate_thermal_stress(temp_c: float, rh: float, wind_kmh: float = 0.0) -> ThermalResult:
    """
    temp_c: dry bulb air temperature in Celsius
    rh: relative humidity, 0-100
    wind_kmh: wind speed in km/h (optional refinement)
    """
    t = _to_f(temp_c)
    r = rh * INDIA_HUMIDITY_TOLERANCE_OFFSET

    if t < 80:
        hi_f = 0.5 * (t + 61.0 + (t - 68.0) * 1.2 + r * 0.094)
    else:
        hi_f = (
            -42.379
            + 2.04901523 * t
            + 10.14333127 * r
            - 0.22475541 * t * r
            - 0.00683783 * t * t
            - 0.05481717 * r * r
            + 0.00122874 * t * t * r
            + 0.00085282 * t * r * r
            - 0.00000199 * t * t * r * r
        )

    if wind_kmh > 8:
        hi_f -= min((wind_kmh - 8) * 0.05, 3)

    heat_index_c = round(_to_c(hi_f), 1)

    # Map heat index (C) to a 0-100 score, calibrated against IMD
    # heatwave departure criteria for the Indian plains.
    if heat_index_c <= 30:
        score = max(0.0, (heat_index_c / 30) * 40)
    elif heat_index_c <= 38:
        score = 40 + ((heat_index_c - 30) / 8) * 20
    elif heat_index_c <= 45:
        score = 60 + ((heat_index_c - 38) / 7) * 20
    else:
        score = min(100.0, 80 + ((heat_index_c - 45) / 5) * 20)
    score = round(score, 1)

    if score < 40:
        category: RiskCategory = "Low"
        ndma_tier: NDMATier = "Yellow"
    elif score < 60:
        category = "Moderate"
        ndma_tier = "Yellow"
    elif score < 80:
        category = "High"
        ndma_tier = "Orange"
    else:
        category = "Extreme"
        ndma_tier = "Red"

    return ThermalResult(score=score, heat_index_c=heat_index_c, category=category, ndma_tier=ndma_tier)


if __name__ == "__main__":
    # quick manual sanity check
    for temp, rh in [(28, 40), (36, 55), (42, 35), (47, 20)]:
        result = calculate_thermal_stress(temp, rh)
        print(f"{temp}C / {rh}% RH -> {result}")
